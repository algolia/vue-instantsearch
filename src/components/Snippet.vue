<template>
  <span :class="suit('')" v-html="innerHTML">
  </span>
</template>

<script>
import suit from '../suit';
import { getPropertyByPath } from '../util/object';
import { warn } from '../util/warn';

export default {
  props: {
    hit: {
      type: Object,
      required: true,
    },
    attribute: {
      type: String,
      required: true,
    },
    highlightedTagName: {
      type: String,
      default: 'mark',
    },
  },
  data() {
    return {
      widgetName: 'Snippet',
    };
  },
  methods: {
    suit(block, modifier) {
      return suit(this.widgetName, block, modifier)
    }
  },
  computed: {
    innerHTML() {
      const attributeValue =
        getPropertyByPath(
          this.hit,
          `_snippetResult.${this.attribute}.value`
        ) || '';

      if (process.env.NODE_ENV !== 'production' && attributeValue === '') {
        warn(
          `The "${this.attribute}" attribute might currently not be configured to be snippeted in Algolia.
        See https://www.algolia.com/doc/api-reference/api-parameters/attributesToSnippet/.`
        );
      }
      return attributeValue
        .replace(
          new RegExp('<em>', 'g'),
          `<${this.highlightedTagName} class="${this.suit('highlighted')}">`
        )
        .replace(new RegExp('</em>', 'g'), `</${this.highlightedTagName}>`);
    },
  },
};
</script>
