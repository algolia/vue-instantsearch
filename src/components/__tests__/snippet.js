import { mount } from '@vue/test-utils';
import Snippet from '../Snippet.vue';

afterEach(() => {
  process.env.NODE_ENV = 'test';
});

test('renders proper HTML', () => {
  const hit = {
    _snippetResult: {
      attr: {
        value: `con<em>ten</em>t`,
      },
    },
  };

  const wrapper = mount(Snippet, {
    propsData: {
      attribute: 'attr',
      hit,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('renders proper HTML with highlightTagName', () => {
  const hit = {
    _snippetResult: {
      attr: {
        value: `con<em>ten</em>t`,
      },
    },
  };

  const wrapper = mount(Snippet, {
    propsData: {
      attribute: 'attr',
      highlightedTagName: 'marquee',
      hit,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});

test('should render an empty string in production if attribute is not snippeted', () => {
  process.env.NODE_ENV = 'production';
  const hit = {
    _snippetResult: {},
  };
  global.console.warn = jest.fn();

  const wrapper = mount(Snippet, {
    propsData: {
      attribute: 'attr',
      hit,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
  expect(global.console.warn).not.toHaveBeenCalled();
});

test('should warn when not in production if attribute is not snippeted', () => {
  global.console.warn = jest.fn();

  const hit = {
    _snippetResult: {},
  };
  mount(Snippet, {
    propsData: {
      attribute: 'attr',
      hit,
    },
  });

  expect(global.console.warn).toHaveBeenCalledTimes(1);
});

test('allows usage of dot delimited path to access nested attribute', () => {
  const hit = {
    _snippetResult: {
      attr: {
        nested: {
          value: `nested <em>val</em>`,
        },
      },
    },
  };

  const wrapper = mount(Snippet, {
    propsData: {
      attribute: 'attr.nested',
      hit,
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
});
