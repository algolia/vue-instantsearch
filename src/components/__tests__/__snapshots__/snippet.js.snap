// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows usage of dot delimited path to access nested attribute 1`] = `

<span class="ais-Snippet">
  nested
  <mark class="ais-Snippet-highlighted">
    val
  </mark>
</span>

`;

exports[`renders proper HTML 1`] = `

<span class="ais-Snippet">
  con
  <mark class="ais-Snippet-highlighted">
    ten
  </mark>
  t
</span>

`;

exports[`renders proper HTML with highlightTagName 1`] = `

<span class="ais-Snippet">
  con
  <marquee class="ais-Snippet-highlighted">
    ten
  </marquee>
  t
</span>

`;

exports[`should render an empty string in production if attribute is not snippeted 1`] = `

<span class="ais-Snippet">
</span>

`;
